pkg/telemetry/cass_prometheus_resourcer.go

```diff
@@ -0,0 +1,459 @@
+func (cfg CassPrometheusResourcer) NewServiceMonitor() (promapi.ServiceMonitor, error) {
+	// validate the object we're being passed.
+	if cfg.CassandraNamespace == "" || cfg.ServiceMonitorName == "" || cfg.DataCenterName == "" || cfg.ClusterName == "" {
+		return promapi.ServiceMonitor{}, TelemetryConfigIncomplete{}
+	}
+	// Overwrite any CommonLabels the user has asked for if they conflict with the labels essential for the functioning of the operator.
+	var mergedLabels = map[string]string{}
+	for k, v := range cfg.CommonLabels {
+		mergedLabels[k] = v
+	}
+	mustLabels := cfg.mustLabels()
+	for k, v := range mustLabels {
+		mergedLabels[k] = v
+	}
+	var endpointHolder promapi.ServiceMonitor
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	_, _, err := decode([]byte(endpointString), nil, &endpointHolder)
+	if err != nil {
+		return promapi.ServiceMonitor{}, err
+	}
+	sm := promapi.ServiceMonitor{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ServiceMonitor",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cfg.ServiceMonitorName,
+			Namespace: cfg.CassandraNamespace,
+			Labels:    mergedLabels,
+		},
+		Spec: promapi.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app.kubernetes.io/managed-by": "cass-operator",
+					//It appears that this label is always set true by cass-operator, so I don't think filtering on it adds any value and we should look to deprecate.
+					//"cassandra.datastax.com/prom-metrics": "true",
+				},
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{Key: "cassandra.datastax.com/cluster", Operator: "In", Values: []string{cfg.ClusterName}},
+					{Key: "cassandra.datastax.com/datacenter", Operator: "In", Values: []string{cfg.DataCenterName}},
+				},
+			},
+			NamespaceSelector: promapi.NamespaceSelector{
+				MatchNames: []string{cfg.CassandraNamespace},
+			},
+			TargetLabels: []string{
+				"cassandra.datastax.com/cluster",
+				"cassandra.datastax.com/datacenter",
+			},
+			Endpoints: endpointHolder.Spec.Endpoints,
+		},
+	}
+	utils.AddHashAnnotation(&sm, k8ssandraapi.ResourceHashAnnotation)
+	return sm, nil
+}
+
+// GetCassandraPromSMName gets the name for our ServiceMonitors based on
+func GetCassandraPromSMName(cfg CassTelemetryResourcer) string {
+	return strings.Join([]string{cfg.ClusterName, cfg.DataCenterName, "prom-servicemonitor"}, "-")
+}
+
+// CreateResources executes the creation of the desired Prometheus resources on the cluster.
+func (cfg CassPrometheusResourcer) UpdateResources(ctx context.Context, client runtimeclient.Client, owner *k8ssandraapi.K8ssandraCluster) error {
+	desiredSM, err := cfg.NewServiceMonitor()
+	if err != nil {
+		return err
+	}
+	cfg.CassTelemetryResourcer.Logger.Info("checking whether Prometheus ServiceMonitor for Cassandra already exists")
+	// Logic to handle case where SM does not exist.
+	var actualSM promapi.ServiceMonitor
```

Comment:
It would be simpler IMO to use pointers for variables like `desiredSM` and `actualSM`.

Reply:
Sure, I don't see why not! Done.